kernel/Base/LazyCached.h

```diff
@@ -0,0 +1,196 @@
+     *  Note: For the computation, the value is initialized to the default.
+     *  \param computation The computation to update the value
+     */
+    explicit LazyCached(std::function<void(T&)> computation)
+        : compute_(computation), hasValue_(false), value_() {}
+
+    /**
+     * Utility constructor using a non static member function as update
+     * computation.
+     * @tparam C The type of the class
+     * @param instance The instance to call the member function on
+     * @param func Pointer to the member function.
+     */
+    template <typename C>
+    LazyCached(C* instance, void (C::*func)(T&))
+        : LazyCached(std::bind(func, instance, std::placeholders::_1)){};
+
+    /// Get the cached value
+    const T& get() {
+        if (!hasValue_) {
+            compute_(value_);
+            hasValue_ = true;
+        }
+        return value_;
+    }
+
+    /// Reset the cache to a state with no computed value
+    void reset() { hasValue_ = false; }
+
+   private:
+    /// Function to update the value
+    ///
+    /// Design note: the signature is chosen so that the value can be updated
+    /// in place. This can for example be used in conjunction with a vector
+    /// as value to prevent allocating new vectors each time.
+    std::function<void(T&)> compute_;
+    /// Whether or not a value has been computed
+    bool hasValue_{};
+    T value_;
+};
+
+/**
+ * The equivalent to a vector of LazyCached instances with a single computation.
+ *
+ * Note: Just like LazyCached this class is non thread-safe.
+ *
+ * @tparam T The type of value that is actually stored
+ */
+template <typename T>
+class LazyVectorCached {
+    using Entry = std::pair<bool, T>;
+
+   public:
+    /**
+     * Default constructor, uses a comptuation that does not update the value.
+     */
+    LazyVectorCached() : compute_([](T&, std::size_t) {}), values_(){};
+
+    /**
+     * Create a cache vector backed by a computation that is used to compute the
+     * entries of the vector on demand.
+     *
+     * Note the computation will be run only the first time the value for that
+     * index is requested, and for each time after the first reset. This needs
+     * to be taken into account for any side effects the computation may have.
+     *
+     * The first computation for a specific index after growing the cache to
+     * include the index, will get a T() as value argument.
+     *
+     * @param computation The computation used to update the value.
+     */
+    explicit LazyVectorCached(std::function<void(T&, std::size_t)> computation)
+        : compute_(computation), values_(){};
+
+    /**
+     * Utility constructor to use a non static member function as computation.
+     * @tparam C The class
+     * @param instance The instance to call the member function on
+     * @param func The member function used to update the cache.
+     */
+    template <typename C>
+    LazyVectorCached(C* instance, void (C::*func)(T&, std::size_t))
```

Comment:
This function signature exists because it allows you to calculate multiple values at once

Reply:
It is so that you don't have to place something like `[this](T& value, std::size_t index){this->computeValues(values, index);` or use `std::bind` with the long placeholders at every point where you want to invoke the constructor.

I tried it, but kept on repeating that initialization, so abstracted into this function.

Reply:
I was not clear enough, I expected a signature like

`
```suggestion
    LazyVectorCached(C* instance, void (C::*func)(T&),std::size_t size)
```

Reply:
Oh ok, that clarifies it. Then no, it does not allow you to compute multiple values at once. It allows a 'vector' of LazyCached values that are backed by a parametrized computation: `vector<LazyCached<T>>`. So the computation gets both the entry that is updated and the index of that entry (which can then affect the computational result).

With the signature you suggested there are two differences:

 1. The size argument at the end, that could equally well be added to the current method. But at the moment it is not used, so also not build.
 2. The computation does not have the index of the entry in the vector that is being updated. The problem with this is that all the entries share the exact same computation. So there is no natural way to for entries at different position to be different.

The original idea of calculating multiple values at once, I would expect to be the type `LazyCached<vector<T>>`. Note the difference in order with `vector<LazyCached<T>>`, which is the equivalent of LazyVectorCached. In the former the vector as a whole is lazy, while for the latter the entries are lazy.

Reply:
okay thanks for the explanation